sfge/src/main/java/com/salesforce/graph/ops/expander/PathExpansionRegistry.java

```diff
@@ -0,0 +1,151 @@
+package com.salesforce.graph.ops.expander;
+
+import com.google.common.collect.ImmutableMap;
+import com.salesforce.graph.ApexPath;
+import com.salesforce.graph.ops.registry.Indexable;
+import com.salesforce.graph.ops.registry.Registry;
+import com.salesforce.graph.ops.registry.RegistryData;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversalSource;
+
+/**
+ * Optimization to maintain lower heap space usage while expanding paths. See {@link Registry} for
+ * information on how the optimization works.
+ *
+ * <p>This registry implementation holds data for {@link ApexPathCollapser}, {@link
+ * ApexPathExpander}, and {@link ForkEvent} instances. Collections that need to hold instances of
+ * these types, can hold their {@link Long} Ids instead, and use the corresponding <code>lookup()
+ * </code> method to fetch the instance from registry.
+ *
+ * <p>{@link PathExpansionRegistry} uses instances of {@link RegistryData} rather than {@link
+ * ThreadLocal} so that these hashmaps only grow as needed per path expansion of an ApexPath. Once
+ * cleared, all heap usage will be released back. This makes sure:
+ *
+ * <ul>
+ *   <li>1. Registry data lives for a shorted duration of time and doesn't move to Old Gen of heap
+ *       space until later. This allows earlier clean up from heap space when they are not used.
+ *   <li>2. Internally-used HashMaps can start with smaller sizes and can grow as needed, as opposed
+ *       to starting with a much larger HashMap that was needed in a previous path expansion.
+ * </ul>
+ *
+ * <p>Create a new instance of this registry when a path expansion for an ApexPath is initiated at
+ * {@link ApexPathExpanderUtil#expand(GraphTraversalSource, ApexPath, ApexPathExpanderConfig)}.
+ * Clear the registry when the operation completes to avoid accidental memory leaks.
+ */
+public class PathExpansionRegistry extends Registry {
+
+    private static final Map<Class<? extends Indexable>, Supplier> REGISTRY_SUPPLIER =
+            ImmutableMap.of(
+                    ApexPathCollapser.class, () -> new RegistryData<ApexPathCollapser>(),
+                    ForkEvent.class, () -> new RegistryData<ForkEvent>(),
+                    ApexPathExpander.class, () -> new RegistryData<ApexPathExpander>());
```

Comment:
@jfeingold35 I made this change to make RegistryData non-abstract, but the information on YourKit profiler looks less helpful now. I'm going back to make this a class extension to keep performance analysis easier. Will leave a comment on why this is preferred.